### PR TITLE
Erdpy cli contract verify

### PR DESCRIPTION
**Pull request by popenta**
Added a new endpoint to erdpy-contract-cli called `verify` that is **experimental** and interacts with a service that is still work in progress.

Implemented a mock server that runs on `localhost:7777`.
Also added a test.

After the server is started the **erdpy contract verify** command can be used and should return a dummy response.

### CODE REVIEW

File: erdpy/contract_verification.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        source_code = PackagedSourceCode.from_dict(source_code).to_dict()
+    elif project_directory:
+        source_code = PackagedSourceCode.from_folder(project_directory).to_dict()
+    else:
+        raise NotImplementedError()
+
+    secret_key = bytes.fromhex(owner.secret_key)
+    signing_key: Any = SigningKey(secret_key)
+
+    response = requests.post(f"{verifier_url}/initialise", data={"contract": contract.bech32()})
```
Comment (andreibancioiu) [inline erdpy/contract_verification.py:53]: `initialize` (US English / default).

File: erdpy/tests/local_verify_server.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+class HTTP(BaseHTTPRequestHandler):
+    def do_POST(self):
+        self.send_response(200)
+        self.send_header("Content-type", "application/json")
+        self.end_headers()
+
+        if self.path == "/initialise":
+            response = {'token': 7890}
+            print('dasdada')
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy/tests/local_verify_server.py:16]: Deleted! I forgot about them :)

File: erdpy/cli_contracts.py [hunk 3]
```diff
@@ -121,6 +122,33 @@ def setup_parser(args: List[str], subparsers: Any) -> Any:
...
+        "--packaged-src", help="JSON file containing the source code of the contract"
+    )
+
+    _add_contract_arg(sub)
+    sub.add_argument(
+        "--verifier-url",
+        required=True,
+        help="the url of the service that validates the contract",
+    )
+    sub.add_argument("--docker-tag", required=True, help="the tag of the docker image")
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy/cli_contracts.py:148]: Added the example in the description

File: erdpy/cli_contracts.py [hunk 4]
```diff
@@ -303,6 +331,24 @@ def _prepare_sender(args: Any) -> Account:
     return sender
 
 
+def _prepare_owner(args: Any) -> Account:
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy/cli_contracts.py:334]: Renamed it to `_prepare_sender`

File: erdpy/contract_verification.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        source_code: Dict[str, Any],
+        signature: Any,
+        docker_tag: str,
+    ) -> None:
+        self.contract = contract
+        self.source_code = source_code
+        self.signature = signature
+        self.docker_tag = docker_tag
+
+    def to_dictionary(self) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy/contract_verification.py:23]: Rewrote it as above.

File: erdpy/errors.py [hunk 6]
```diff
@@ -193,3 +193,8 @@ def __init__(self, message: str):
 class LedgerError(KnownError):
     def __init__(self, message: str):
         super().__init__("Ledger error: " + message)
+
+
+class ContractSourceCodeNotProvided(KnownError):
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy/errors.py:198]: I think it was used in the initial implementation. I've deleted it.

File: erdpy/contract_verification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy/contract_verification.py:53]: Extracted it in a separate function called `get_verification_token`

File: erdpy/contract_verification.py [hunk 7]
```diff
@@ -13,22 +13,20 @@ def __init__(
...
-        contract_verification_request["source_code"] = self.source_code
-        contract_verification_request["signature"] = self.signature
-        contract_verification_request["docker_tag"] = self.docker_tag
-
-        return contract_verification_request
+        return {
+            "contract": self.contract,
+            "source_code": self.source_code,
+            "signature": self.signature,
+            "docker_tag": self.docker_image
```
Comment (andreibancioiu) [inline erdpy/contract_verification.py:28]: The dictionary key can be `dockerImage` (JSON-friendly). Also, `sourceCode` (cammel case).

Comment (popenta) [reply erdpy/contract_verification.py]: Done

**Review by andreibancioiu — APPROVED** [review]

File: erdpy/cli_contracts.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy/cli_contracts.py]: Actually, after renaming it I saw that there is already a method called `_prepare_sender` so I'll rename the method.

**Review by andreibancioiu — DISMISSED** [review]

File: erdpy/tests/test_cli_contracts.sh [hunk 8]
```diff
@@ -120,6 +120,18 @@ testCleanContracts() {
...
+testVerifyContract(){
+    echo "testVerifyContract"
+
+    nohup python3 local_verify_server.py >/dev/null 2>&1 &
+    sleep 1
+
+    curl localhost:7777/initialise -X POST
+    curl localhost:7777/verify -X POST
+
+    pkill -f local_verify_server.py
```
Comment (claudiu725) [inline erdpy/tests/test_cli_contracts.sh:132]: There should be a call to `${ERDPY} contract verify` somewhere in this test, and then a test that the response matches.

**Review by claudiu725 — CHANGES_REQUESTED** [review]

**Review by andreibancioiu — DISMISSED** [review]

File: erdpy/tests/test_cli_contracts.sh [hunk 9]
```diff
@@ -120,6 +120,31 @@ testCleanContracts() {
...
 
+testVerifyContract(){
+    echo "testVerifyContract"
+
+    nohup python3 local_verify_server.py >/dev/null 2>&1 &
+    sleep 1
+
+    query_response=$(curl -s localhost:7777/verify -X POST)
+
+    command_response=$(${ERDPY} contract verify erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th \
```
Comment (andreibancioiu) [inline erdpy/tests/test_cli_contracts.sh:131]: Perhaps we can use an example address that is a smart contract. E.g. one from here:

https://devnet-explorer.elrond.com/accounts/erd1testnlersh4z0wsv8kjx39me4rmnvjkwu8dsaea7ukdvvc9z396qykv7z7/contracts

**Review by andreibancioiu — APPROVED** [review]

**Review by claudiu725 — APPROVED** [review]

**Pull request merged**